SonarQube.Common/AnalysisProperties/CmdLineArgPropertyProvider.cs

```diff
@@ -0,0 +1,159 @@
+// </copyright>
+//-----------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace SonarQube.Common
+{
+    /// <summary>
+    /// Handles validating and fetch analysis properties from the command line
+    /// </summary>
+    public class CmdLineArgPropertyProvider : IAnalysisPropertyProvider
+    {
+        public const string DynamicPropertyArgumentId = "dynamic.property";
+
+        public static readonly ArgumentDescriptor Descriptor = new ArgumentDescriptor(
+            id: DynamicPropertyArgumentId, prefixes: new string[] { "/p:" }, required: false, allowMultiple: true, description: Resources.CmdLine_ArgDescription_DynamicProperty);
+
+        private readonly IEnumerable<Property> properties;
+           
+        #region Public methods
+
+        /// <summary>
+        /// Attempts to construct and return a provider that uses analysis properties provided on the command line
+        /// </summary>
+        /// <param name="commandLineArguments">List of command line arguments (optional)</param>
+        /// <returns>False if errors occurred when constructing the provider, otherwise true</returns>
+        /// <remarks>If no properties were provided on the command line then an empty provider will be returned</remarks>
+        public static bool TryCreateProvider(IEnumerable<ArgumentInstance> commandLineArguments, ILogger logger, out CmdLineArgPropertyProvider provider)
+        {
+            if (commandLineArguments == null)
+            {
+                throw new ArgumentNullException("commandLineArguments");
+            }
+            if (logger == null)
+            {
+                throw new ArgumentNullException("logger");
+            }
+
+            IEnumerable<Property> validProperties;
+            if (TryGetAnalysisProperties(commandLineArguments, logger, out validProperties))
+            {
+                provider = new CmdLineArgPropertyProvider(validProperties);
+                return true;
+            }
+
+            provider = null;
+            return false;
+        }
+
+        #endregion
+
+
+        #region IAnalysisPropertyProvider interface
+
+        public bool TryGetProperty(string key, out Property property)
+        {
+            return Property.TryGetProperty(key, this.properties, out property);
+        }
+           
+        public IEnumerable<Property> GetAllProperties()
+        {
+            return this.properties ?? Enumerable.Empty<Property>();
+        }
+
+        #endregion
+
+        #region Private methods
+
+        private CmdLineArgPropertyProvider(IEnumerable<Property> properties)
+        {
+            if (properties == null)
+            {
+                throw new ArgumentNullException("properties");
+            }
+            this.properties = properties;
+        }
+
+        #endregion
+
+        #region Analysis properties handling
```

Comment:
The rest of the code in this file was copied from the ArgumentProcessor class and tweaked to use the new "Property" instead of "AnalysisSetting". "Property" and "AnalysisSetting" are very similar (so much so that I originally used AnalysisSetting). However, the XML serialization of AnalysisSetting isn't what we wanted. Also, it's easier to keep track of the "old" and "newer" code during refactoring if they don't both use AnalysisSetting.
We might consolidate them into a single class again later.